### PR TITLE
ASoC: SOF: DMIC: Fixed length IPC

### PR DESCRIPTION
**Pull request by singalsu**

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -2881,18 +2881,13 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
-	u32 size;
+	u32 size = sizeof(*config);
 	int ret, j;
 
-	/*
-	 * config is only used for the common params in dmic_params structure
-	 * that does not include the PDM controller config array
-	 * Set the common params to 0.
-	 */
+	/* Ensure the entrire DMIC config struct is zeros */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: typo: entire

**Comment by singalsu** [comment]
The similar PR for firmware would increase ABI minor to 15. Since the ABI minor for current kernel is behind by one this would increase ABI minor to 14 that is not consistent. Probably the feature from firmware should be merged first to get to version 14 and have 15 on both sides with this merged.

I've tested this with all combinations of new/new, new/old, old/new for kernel and FW on UP2+DMIC kit and no issues seen.

File: include/sound/sof/dai-intel.h [hunk 2]
```diff
@@ -184,8 +187,8 @@ struct sof_ipc_dai_dmic_params {
 	/* reserved for future use */
 	uint32_t reserved[5];
 
-	/**< variable number of pdm controller config */
-	struct sof_ipc_dai_dmic_pdm_ctrl pdm[0];
+	/**< Fixed number of pdm controller config */
```
Comment (plbossart) [inline include/sound/sof/dai-intel.h]: it's not really fixed, it's rather a maximum. the number of active pcm controllers is still variable, right?

File: include/sound/sof/dai-intel.h [hunk 3]
```diff
@@ -49,6 +49,9 @@
 /* bclk idle */
 #define SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_IDLE_HIGH	BIT(5)
 
+/* DMIC max. four controllers for eight microphone channels */
```
Comment (plbossart) [inline include/sound/sof/dai-intel.h:52]: I thought it was 2 and 4 channels respectively? Which device supports 8  channels?

File: include/sound/sof/dai-intel.h [hunk 4]
```diff
@@ -49,6 +49,9 @@
...
+#define SOF_DAI_INTEL_DMIC_NUM_PDM			4
```
Comment (plbossart) [inline include/sound/sof/dai-intel.h]: NUM_DMIC_CTRL? that would avoid ambiguities with channels?

File: include/sound/sof/dai-intel.h (hunk 2 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: Yep, IPC is for max configuration. The active controllers count is still passed in the header and defines how much of this struct is filled. Should the comment text mention this?

File: include/sound/sof/dai-intel.h (hunk 3 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h:52]: Only Sue Creek of SOF supported platforms supports 8 channels. The platforms for notebooks are 4 channels. If Sue Creek is no more supported the macro value could be 2.

File: include/sound/sof/dai-intel.h (hunk 4 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: That sounds like a good proposal!

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (singalsu) [reply sound/soc/sof/topology.c]: Done

File: include/sound/sof/dai-intel.h (hunk 4 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: Done

File: include/sound/sof/dai-intel.h (hunk 2 above)
Comment (ranj063) [reply include/sound/sof/dai-intel.h]: @singalsu I think so, yes

File: include/sound/sof/dai-intel.h (hunk 3 above)
Comment (plbossart) [reply include/sound/sof/dai-intel.h:52]: ah, ok. that's fine then.

File: include/sound/sof/dai-intel.h (hunk 2 above)
Comment (plbossart) [reply include/sound/sof/dai-intel.h]: Maybe instead add a comment on line 181 

	uint32_t num_pdm_active; /**< Number of active pdm controllers. Range is 1..SOF_DAI_INTEL_DMIC_NUM_CTRL */

File: include/uapi/sound/sof/abi.h [hunk 5]
```diff
@@ -26,7 +26,7 @@
 
 /* SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 13
+#define SOF_ABI_MINOR 14
```
Comment (lyakh) [inline include/uapi/sound/sof/abi.h]: I seem to remember a PR that was bumping up the minor number to 15. I see 14 already used in the firmware. Maybe we need a more strict and explicit ABI allocation list somewhere. Or haven't we had one? Maybe add a feature on GH with a list of ABI numbers with a short description of a reason for the change and a link to the firmware and kernel PRs?

File: sound/soc/sof/topology.c [hunk 6]
```diff
@@ -2881,18 +2881,13 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_soc_tplg_private *private = &cfg->priv;
-	struct sof_ipc_dai_config *ipc_config;
 	struct sof_ipc_reply reply;
 	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
 	struct sof_ipc_fw_version *v = &ready->version;
-	u32 size;
+	u32 size = sizeof(*config);
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: `size` is indeed used to assign to `config->hdr.size` which is of type `uint32_t` and it's used as an argument for `sof_set_dai_config()` which is of the same type, but the latter is actually isn't a good idea and it would be good to change it. So, it isn't completely wrong to use `u32` here, but I would actually incline to `size_t`

File: sound/soc/sof/topology.c [hunk 7]
```diff
@@ -2941,44 +2919,42 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
-			ipc_config->dmic.pdm[j].skew);
+			config->dmic.pdm[j].id,
+			config->dmic.pdm[j].clk_edge,
+			config->dmic.pdm[j].skew);
 	}
 
 	if (SOF_ABI_VER(v->major, v->minor, v->micro) < SOF_ABI_VER(3, 0, 1)) {
 		/* this takes care of backwards compatible handling of fifo_bits_b */
-		ipc_config->dmic.reserved_2 = ipc_config->dmic.fifo_bits;
+		config->dmic.reserved_2 = config->dmic.fifo_bits;
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: ouch... Ok, this isn't introduced by this patch, but this doesn't look good. We have a field that is called "reserved" so anyone might come up with an idea to begin using it or to remove it. If it is used it shouldn't be called "reserved." What am I missing?

File: include/sound/sof/dai-intel.h (hunk 2 above)
Comment (kv2019i) [reply include/sound/sof/dai-intel.h]: @singalsu I'd change both. On L190 state: "pdm controller config" and L181 put the word @plbossart has above.

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -2905,33 +2900,16 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
-	memcpy(ipc_config, config, sizeof(*config));
-
 	/*
 	 * alloc memory for private member
 	 * Used to track the pdm config array index currently being parsed
 	 */
 	sdev->private = kzalloc(sizeof(u32), GFP_KERNEL);
-	if (!sdev->private) {
-		kfree(ipc_config);
+	if (!sdev->private)
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2918]: this `sdev->private` is really weird. DMIC uses it to allocate one integer for topology parsing... Whereas iMX8 uses it for its private data. So, iMX8 cannot use DMICs? DMIC *really* shouldn't use a device-global "private" field for its own purpose. Could you add a patch to this PR, adding a member to `struct snd_sof_dev` like
```
 	struct list_head route_list;
 	struct snd_soc_component *component;
 	u32 enabled_cores_mask; /* keep track of enabled cores */
+	unsigned int pdm_index; /* PDM configuration array index */
 
 	/* FW configuration */
 	struct sof_ipc_window *info_window;
```

File: sound/soc/sof/topology.c (hunk 7 above)
Comment (lyakh) [reply sound/soc/sof/topology.c]: in fact, since this PR already changes the ABI, maybe add a patch too it to properly name that field?

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (kv2019i) [inline include/uapi/sound/sof/abi.h]: There's some confusion now with ABI versions https://github.com/thesofproject/linux/pull/1929/
FYI @xiulipan

File: include/sound/sof/dai-intel.h (hunk 2 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: @kv2019i Yep, but I feel comment "/**< PDM controllers configuration */" is better.

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (singalsu) [reply include/uapi/sound/sof/abi.h]: Would be good, but to me it sounds like a task for maintainers and not required for this PR.

I think I'l lookup with git blame for abi.h what was the old purpose of reserved_2 before it got deprecated (below). Such list would be useful for such needs too!

File: sound/soc/sof/topology.c (hunk 6 above)
Comment (singalsu) [reply sound/soc/sof/topology.c]: Yep, I was about to to change it myself too but went for deltas minimization. With your permit I'll change!

File: sound/soc/sof/topology.c (hunk 7 above)
Comment (singalsu) [reply sound/soc/sof/topology.c]: OK, I'll put that to another commit.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (singalsu) [reply sound/soc/sof/topology.c:2918]: @ranj063 Can you comment this proposal? Personally I don't think that such change belongs to this PR. It could be handled separately.

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (singalsu) [reply include/uapi/sound/sof/abi.h]: I think I'll change this patch to 15.. to align with firmware PR.

Comment (paulstelian97) [reply include/uapi/sound/sof/abi.h]: @singalsu Is that the consensus on what it should be? I noticed Liam put it in the ABI change classifier.

Comment (singalsu) [reply include/uapi/sound/sof/abi.h]: No, obviously it can't be 14. I'll change this PR again when it's known @lgirdwood @plbossart .

File: include/sound/sof/dai-intel.h (hunk 2 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: Here's new proposal version, is this part OK?

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (singalsu) [reply include/uapi/sound/sof/abi.h]: Now proposing 15 but then maybe it needs to be 16. Waiting for instructions.

File: sound/soc/sof/topology.c (hunk 6 above)
Comment (singalsu) [reply sound/soc/sof/topology.c]: Done

File: sound/soc/sof/topology.c (hunk 7 above)
Comment (singalsu) [reply sound/soc/sof/topology.c]: Done

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (singalsu) [reply include/uapi/sound/sof/abi.h]: At 15 now but in above comment by @lyakh there's already PR with proposal for same so maybe this needs to be 16 for both kernel and FW. Waiting for guidance!

**Comment by singalsu** [comment]
I've addressed the comments in my way except the sdev->private and ABI minor version.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:2918]: @singalsu @lyakh this is beyond the scope of this PR. Besides @juimonen  PR takes care of removing this anyway

**Comment by singalsu** [comment]
On FW side review @lgirdwood said this should have ABI minor version 14 so my guess was wrong. I'll wait for other comments before I update this.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:2918]: @ranj063 @juimonen what PR is that?

File: sound/soc/sof/topology.c [hunk 9]
```diff
@@ -2947,9 +2947,12 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
 	}
 
+	/*
+	 * this takes care of backwards compatible handling of fifo_bits_b.
+	 * It is deprecated since 3.0.1.
+	 */
 	if (SOF_ABI_VER(v->major, v->minor, v->micro) < SOF_ABI_VER(3, 0, 1)) {
-		/* this takes care of backwards compatible handling of fifo_bits_b */
-		config->dmic.reserved_2 = config->dmic.fifo_bits;
+		config->dmic.fifo_bits_b = config->dmic.fifo_bits;
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2965]: some people prefer to use curly braces if there is more than one *line* in the block, even though it's just one command, e.g.
```
if (ret < 0) {
	/* error */
	return ret;
}
```
or
```
if (ret < 0) {
	dev_err(dev, "error %d\n",
			ret);
}
```
but here we have one command on one line... Do we still need braces?..

Comment (paulstelian97) [reply sound/soc/sof/topology.c:2965]: @lyakh I'd say it would be pretty confusing and error-prone given the comment. So I for one would recommend keeping the braces as long as the comment exists.

Comment (singalsu) [reply sound/soc/sof/topology.c:2965]: Yep, I'll change. Maybe those were used because of the comment.

**Comment by singalsu** [comment]
> Looks good now @singalsu
> 
> Not sure who should do this, but I manually added this PR to the ABI classifier as well:
> https://github.com/orgs/thesofproject/projects/2?card_filter_query=dmic

Thanks! I didn't know about that. I'll learn how to do it.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (singalsu) [reply sound/soc/sof/topology.c:2918]: (This is resolved but I'll leave open to let continue the discussion)

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (singalsu) [reply include/uapi/sound/sof/abi.h]: Now at 14

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (singalsu) [reply sound/soc/sof/topology.c:2965]: Done!

**Comment by singalsu** [comment]
I just pushed new version with extra if statement brackets removed and changed ABI minor back to 14.

File: include/uapi/sound/sof/abi.h (hunk 5 above)
Comment (kv2019i) [reply include/uapi/sound/sof/abi.h]: it's 14 as per https://github.com/orgs/thesofproject/projects/2

**Comment by singalsu** [comment]
> Thanks @singalsu , looks good now! Merging will pend on approval for sof#2610

Thanks, I'll work with it next.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:2918]: @lyakh it is https://github.com/thesofproject/linux/pull/1766

File: include/sound/sof/dai-intel.h (hunk 3 above)
Comment (plbossart) [inline include/sound/sof/dai-intel.h:52]: nit on commit title; Can we use ASoC: SOF: Intel: to show this is not a core change but an Intel specific one?

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:2918]: good feedback @lyakh let's handle this separately.

File: include/sound/sof/dai-intel.h [hunk 10]
```diff
@@ -173,7 +173,7 @@ struct sof_ipc_dai_dmic_params {
 	uint32_t fifo_fs;	/**< FIFO sample rate in Hz (8000..96000) */
 	uint32_t reserved_1;	/**< Reserved */
 	uint16_t fifo_bits;	/**< FIFO word length (16 or 32) */
-	uint16_t reserved_2;	/**< Reserved */
+	uint16_t fifo_bits_b;	/**< Deprecated since 3.0.1 */
```
Comment (plbossart) [inline include/sound/sof/dai-intel.h]: is 3.0.1 the ABI level? maybe clarify is this is firmware version or ABI to remove any ambiguity.

File: sound/soc/sof/topology.c [hunk 11]
```diff
@@ -2947,10 +2947,12 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
 			config->dmic.pdm[j].skew);
 	}
 
-	if (SOF_ABI_VER(v->major, v->minor, v->micro) < SOF_ABI_VER(3, 0, 1)) {
-		/* this takes care of backwards compatible handling of fifo_bits_b */
-		config->dmic.reserved_2 = config->dmic.fifo_bits;
-	}
+	/*
+	 * this takes care of backwards compatible handling of fifo_bits_b.
+	 * It is deprecated since 3.0.1.
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: ABI version 3.0.1 to align with code.

File: include/sound/sof/dai-intel.h [hunk 12]
```diff
@@ -173,7 +173,7 @@ struct sof_ipc_dai_dmic_params {
 	uint32_t fifo_fs;	/**< FIFO sample rate in Hz (8000..96000) */
```
Comment (singalsu) [inline include/sound/sof/dai-intel.h:173]: Sorry, I made a mistake with commit title. This needs to be ASoC: SOF: Intel:

File: include/sound/sof/dai-intel.h (hunk 3 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h:52]: This header file change is for Intel only but the change in topology to remove the capability for variable length array is common. I didn't originally split these because git tree with only one of the changes would cause failed functionality.

Should I do the split or just say this is Intel specific change. No other than Intel SOF platforms today use this load code for DMIC DAI.

File: include/sound/sof/dai-intel.h (hunk 10 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: Yep. It's for loaded firmware ABI level that is got via sdev->fw_ready->version.

File: include/sound/sof/dai-intel.h (hunk 3 above)
Comment (plbossart) [reply include/sound/sof/dai-intel.h:52]: it's not really common, the dmic load is only for Intel. it's fine to keep the code as is. just change the commit title

Comment (singalsu) [reply include/sound/sof/dai-intel.h:52]: I just tried my luck by just changing title of second commit with less effort option. Please vote if OK or not!

File: include/sound/sof/dai-intel.h (hunk 10 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h]: Done

File: include/sound/sof/dai-intel.h (hunk 12 above)
Comment (singalsu) [reply include/sound/sof/dai-intel.h:173]: Done

File: include/sound/sof/dai-intel.h (hunk 3 above)
Comment (plbossart) [reply include/sound/sof/dai-intel.h:52]: it's fine

**Comment by singalsu** [comment]
Thanks for quick review work to everyone. The firmware proposal PR is now updated to match this one.

**Comment by kv2019i** [comment]
Just waiting on https://github.com/thesofproject/sof/pull/2610 now..
UPDATE: sof#2610 now approved. I'll let FW merge first and then I'll merge this.

**Comment by singalsu** [comment]
This leaves abi.h at minor 15. No other changes into this PR.

File: sound/soc/sof/topology.c [hunk 13]
```diff
@@ -2915,33 +2910,16 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
-
-	/* copy the common dai config and dmic params */
-	memcpy(ipc_config, config, sizeof(*config));
-
 	/*
 	 * alloc memory for private member
 	 * Used to track the pdm config array index currently being parsed
 	 */
 	sdev->private = kzalloc(sizeof(u32), GFP_KERNEL);
-	if (!sdev->private) {
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2938]: Do I understand correctly that this now conflicts with #1766? But the conflict seems to be easy to resolve.

File: include/sound/sof/dai-intel.h [hunk 14]
```diff
@@ -173,7 +173,7 @@ struct sof_ipc_dai_dmic_params {
...
 	uint32_t reserved_1;	/**< Reserved */
 	uint16_t fifo_bits;	/**< FIFO word length (16 or 32) */
-	uint16_t reserved_2;	/**< Reserved */
+	uint16_t fifo_bits_b;	/**< Deprecated since firmware ABI 3.0.1 */
```
Comment (lyakh) [inline include/sound/sof/dai-intel.h:176]: I'm wondering how we're going to handle such deprecation? Do we set a grace period of, say, a year and then remove this backward compatibility and then claim the field as "reserved?"

File: sound/soc/sof/topology.c (hunk 13 above)
Comment (kv2019i) [reply sound/soc/sof/topology.c:2938]: @juimonen FYI, Seppo is out this week, so maybe let's put this #1924 in first, and you'll update the other one?

File: include/sound/sof/dai-intel.h (hunk 14 above)
Comment (kv2019i) [reply include/sound/sof/dai-intel.h:176]: @lyakh I don't think you really ever get back reserved bits via deprecation unless you end up in a situation where you absolutely have to. I.e. rare enough event, so better worry about the process when the time comes.

File: sound/soc/sof/topology.c (hunk 13 above)
Comment (lyakh) [reply sound/soc/sof/topology.c:2938]: @kv2019i you want me to update it? Sure can do, but so far I'm not aware of many cases of such take-overs. Is it urgent? Or was it just a typo?

File: include/sound/sof/dai-intel.h [hunk 15]
```diff
@@ -49,6 +49,9 @@
 /* bclk idle */
 #define SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_IDLE_HIGH	BIT(5)
 
+/* DMIC max. four controllers for eight microphone channels */
+#define SOF_DAI_INTEL_DMIC_NUM_CTRL			4
```
Comment (lyakh) [inline include/sound/sof/dai-intel.h:53]: sorry for chiming in so late with this. This can be made a separate PR: should we add a check of `num_pdm_active <= SOF_DAI_INTEL_DMIC_NUM_CTRL`? I think it would be good to tighten up our topology consistency checks.

File: sound/soc/sof/topology.c [hunk 16]
```diff
@@ -2891,18 +2891,13 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
-	u32 size;
+	size_t size = sizeof(*config);
 	int ret, j;
 
-	/*
-	 * config is only used for the common params in dmic_params structure
-	 * that does not include the PDM controller config array
-	 * Set the common params to 0.
-	 */
+	/* Ensure the entire DMIC config struct is zeros */
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2900]: This comment could be just drop, it doesn't seem to add a lot of information. In fact, looking at it: the below memset() is completely redundant. The caller has already 0-initialised the whole `config`. @singalsu could you make an additional PR to remove this too?

**Comment by kv2019i** [comment]
Thanks @lyakh . Let' s merge this version if the FW side progresses this week, and handle updates in separate PR.

**Comment by kv2019i** [comment]
FW part merged, merging this.
@singalsu Please check the additional comments from @lyakh and address in follow-up if still valid.
